src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/Subscription.cs

```diff
@@ -39,51 +41,64 @@ internal sealed class Subscription : ISubscription
+                            await _connection.SendAsync(new DataResultMessage(Id, result), cancellationToken);
+                        }
                     }
                 }
 
-                if (!_cts.IsCancellationRequested)
+                if (!cancellationToken.IsCancellationRequested && !_connection.Closed)
                 {
-                    await _connection.SendAsync(new DataCompleteMessage(Id), _cts.Token);
-                    Completed?.Invoke(this, EventArgs.Empty);
+                    await _connection.SendAsync(new DataCompleteMessage(Id), cancellationToken);
                 }
             }
-            catch (OperationCanceledException) { }
+            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested) { }
             catch (ObjectDisposedException) { }
-            catch (Exception ex)
+            catch (Exception ex) when (!cancellationToken.IsCancellationRequested)
             {
-                if (!_cts.IsCancellationRequested)
+                if (!_connection.Closed)
                 {
-                    IError error =
-                        ErrorBuilder
-                            .New()
-                            .SetException(ex)
-                            .SetCode(ErrorCodes.Execution.TaskProcessingError)
-                            .SetMessage("Unexpected Execution Error")
-                            .Build();
-
-                    IQueryResult result = QueryResultBuilder.CreateError(error);
                     try
                     {
-                        await _connection.SendAsync(new DataResultMessage(Id, result), _cts.Token);
-                    }
-                    finally
-                    {
-                        await _connection.SendAsync(new DataCompleteMessage(Id), _cts.Token);
-                        Completed?.Invoke(this, EventArgs.Empty);
+                        IError error =
+                            ErrorBuilder
+                                .New()
+                                .SetException(ex)
+                                .SetCode(ErrorCodes.Execution.TaskProcessingError)
+                                .SetMessage("Unexpected Execution Error")
+                                .Build();
+
+                        IQueryResult result = QueryResultBuilder.CreateError(error);
+                        try
+                        {
+                            await _connection.SendAsync(new DataResultMessage(Id, result), cancellationToken);
+                        }
+                        finally
+                        {
+                            await _connection.SendAsync(new DataCompleteMessage(Id), cancellationToken);
+                        }
+
                     }
+                    catch { } // suppress all errors, so original exception can be rethrown
                 }
+
+                // original exception should be propagated to upper level in order to be logged correctly at least
```

Comment:
Throwing the exception here will not really help logging. I think we should inject the IDiagnostEvenets and log the error.

Reply:
Well, app developer can always log exception from not-awaited task if add handler to TaskScheduler.UnobservedTaskException, so such exceptions can be logged if desired. In any case I will add IDiagnosticEvents.

Reply:
I have done the events now.

Reply:
The reason for not to rely on `TaskScheduler.UnobservedTaskException` is that. the telemetry api should be the go to to get such insights.